
#' A function for SPIA algorithm analysis.
#'
#' This function allows you to run SPIA method at least 10 times faster using multithreading.
#' @param sessionID This is the sessionID generated by previous "master_LincsDataAnalysis" or "master_diffGenes" functions run.
#' @param gpgene A character string consisting of all significant differentially expressed gene IDs (Entrez IDs) separated by commas.
#' @param path_to_write This necessary parameter specifies the path which user wants to save the 
#'	output files (heatmaps, TreeView files and some other output files) to be saved in.
#' @param fdr False Discovery Rate, if provided set to TRUE.
#' @param cutoff Significance level (alpha).
#' @param species A character string. One of c("Hs", "Mm", "Rn") for Human, Mouse or Rat pathways respectively.
#' @param foldChange A logical value if fold change is provided.
#' @param authors M. Fazel-Najafabadi
#' @keywords SPIA.
#' @export 
#' @examples
#' res <- functionalPathwayAnalysis_SPIA(....)

## Mehdi: SPIA pathway analysis for the gene list

functionalPathwayAnalysis_SPIA <- function(sessionID, gpgene, path_to_write, fdr=FALSE, cutoff = 0.01, species="Hs", foldChange=FALSE)
{
    if (file.exists(paste(path_to_write, "temp_volcano_", sessionID, ".RData", sep=""))) {
	print(paste("loading . . . "))
	load(paste(path_to_write, "temp_volcano_", sessionID, ".RData", sep=""))
	SignatureTable <- SignatureTable[, c("ID_geneid", "Name_GeneSymbol", "Value_LogDiffExp", "Significance_pvalue")]
    } else if (file.exists(paste(path_to_write, "sig_", sessionID, ".xls", sep=""))) {
	print(paste("reading signature . . . "))
	SignatureTable <- read.table(paste(path_to_write, "sig_", sessionID, ".xls", sep=""), header=TRUE, sep="\t", stringsAsFactors=FALSE) # , quote=""
	SignatureTable <- SignatureTable[,-(1:2)]
    } else if (file.exists(paste(path_to_write, "processedSig_", sessionID, ".xls", sep=""))) {
	print(paste("reading processed signature . . . "))
	SignatureTable <- read.table(paste(path_to_write, "processedSig_", sessionID, ".xls", sep=""), header=TRUE, sep="\t", stringsAsFactors=FALSE) # , quote=""
	SignatureTable <- SignatureTable # [,-1]
    } else {
	spiaRes <- data.frame(Nam=NA, ID=NA, pSize=NA, NDE=NA, pNDE=NA, tA=NA, pPERT=NA, pG=NA, pGFdr=NA, Status=NA, KEGGLINK=NA, Remark="File does not exist!", stringsAsFactors=FALSE)
	return(spiaRes)
    }
    if (dim(SignatureTable)[2] == 3) {
	if ("Significance_pvalue" %in% colnames(SignatureTable)) {
	    SignatureTable <- cbind(SignatureTable[,1], NA, SignatureTable[,2:3])
	    colnames(SignatureTable) <- c("GeneID","GeneName","coefficients","Pvals")
	} else {
# 	    colnames(SignatureTable) <- c("GeneID","GeneName","coefficients")
	    spiaRes <- data.frame(Nam=NA, ID=NA, pSize=NA, NDE=NA, pNDE=NA, tA=NA, pPERT=NA, pG=NA, pGFdr=NA, Status=NA, KEGGLINK=NA, Remark="No pValues provided", stringsAsFactors=FALSE)
	    return(spiaRes)
	}
    } else {
	colnames(SignatureTable) <- c("GeneID","GeneName","coefficients","Pvals")
    }
    
#gpgene <- strsplit(as.character(gpgene), split="DELIM")[[1]]
gpgene <- unlist(strsplit(as.character(gpgene), split="DELIM"))

topGenes <- SignatureTable[which(gpgene %in% SignatureTable$GeneID),]
topGenes <- na.omit(topGenes)
if (dim(topGenes)[1] == length(unique(topGenes$GeneID))) rownames(topGenes) <- as.character(topGenes$GeneID) else {
topGenes <- do.call("rbind", lapply(split(topGenes, topGenes$GeneID), function(j) {
		    d <- dim(j)[1]
		    row <- j[order(j$Pvals),][1,]
		    row$Pvals <- row$Pvals * d
		    row
		    }))
rownames(topGenes) <- as.character(topGenes$GeneID)
}
colnames(topGenes) <- c("GeneID", "GeneNames", "log2FoldChange", "pval")
topGenes <- topGenes[order(topGenes$pval),]
  store.wd <- getwd()
  setwd("/opt/raid10/genomics/mehdi/ilincs/SPIA_KEGG_pathways/2017-09-07/")

  
  species <- switch(species, Hs = "hsa", Mm = "mmu", Rn = "rno")
  if (foldChange) {
    sig.geneid <- rownames(topGenes)[abs(topGenes$log2FoldChange) >= log2(cutoff)]
  } else if (fdr) {
    sig.geneid <- rownames(topGenes)[topGenes$padj < cutoff]
  } else {
    sig.geneid <- rownames(topGenes)[topGenes$pval < cutoff]
  }
  sig.geneid <- sig.geneid[!is.na(sig.geneid)]
  
  de.log2FC <- topGenes[sig.geneid, "log2FoldChange"]
  names(de.log2FC) <- sig.geneid
  
  pathway.name <- paste0("KEGG.", species)
  spiaRes <- do.call("runilincsSPIA", list(de=de.log2FC, all=unique(SignatureTable$GeneID), pathwaySetName=pathway.name), envir=.GlobalEnv)#, data.dir=getwd())
  print("SPIA complete")
  KEGGpathways <-  KEGGREST::keggList("pathway", species)
  names(KEGGpathways) <- gsub("path:", "", names(KEGGpathways))
  KEGGpathways <- lapply(strsplit(KEGGpathways, " - "), function(x){paste0(x[-length(x)], collapse=" - ")})
  spiaRes$ID <- names(KEGGpathways)[match(spiaRes$Nam, KEGGpathways)]
  spiaRes <- spiaRes[!is.na(spiaRes$ID), ]
if (dim(spiaRes)[1] > 0) {
a <- Sys.time()
  spiaRes$KEGGLINK <- keggLinks(topGenes=topGenes, pathwayidlist=spiaRes$ID, fdr=fdr, cutoff=cutoff, species=species, foldChange=foldChange, path_to_write=path_to_write, sessionID=sessionID)
##  spiaRes$KEGGLINK <- NA
b <- Sys.time()
print(paste(b-a, " Seconds creating the lincs ****************** !!!"))
spiaRes$Remark <- "Done"
} else { 
spiaRes[1,] <- NA
spiaRes$KEGGLINK <- NA
spiaRes$Remark <- "No significant pathways found"
}
  spiaRes <- spiaRes[order(spiaRes$pG),]
  spiaRes <- spiaRes[, c("Nam", "ID", "pSize", "NDE", "pNDE", "tA", "pPERT", "pG", "pGFdr", "Status", "KEGGLINK", "Remark")]
  colnames(spiaRes) <- c("KEGG pathway name", "KEGG pathway ID", "Genes in Pathway", "DE Genes in Pathway", "ORA pval", "Topology Score", "Top pval", "SPIA pval", "SPIA padj", "Status", "KEGG link", "Remark")
  if (spiaRes$Remark[1] == "Done") {
    write.table(as.data.frame(spiaRes[,1:11]) ,file=paste(path_to_write, "SPIA", sessionID,".xls", sep=""), row.names=FALSE, sep=";", quote=FALSE)
  }
  setwd(store.wd)
  return(spiaRes)
}
