
#' A function for creating KEGG hyperlinks.
#'
#' This function allows you to create many KEGG pathway hyperlincs quikly using several cores, if available.
#' @param pathwayidlist A list of all pathways needed to be used in the analysis.
#' @param topGenes A data.frame object containing geneIDs as row names, log2 fold Changes and pValue or adj_pValues for each gene.
#'	This data.frame should at least have two columns named "log2FoldChange" and "pval"/"padj".
#' @param fdr  A logical value for False Discovery Rate. If needs to be used, set to TRUE.
#' @param cutoff Significance level (alpha).
#' @param species A character string. One of c("Hs", "Mm", "Rn") for Human, Mouse or Rat pathways respectively.
#' @param foldChange A logical value if fold change is provided.
#' @param path_to_write This necessary parameter specifies the path which user wants to save the 
#'	output files (heatmaps, TreeView files and some other output files) to be saved in.
#' @param sessionID This is the sessionID generated by previous "master_LincsDataAnalysis" or "master_diffGenes" functions run.
#' @param authors M. Fazel-Najafabadi
#' @keywords KEGGlinks
#' @export 
#' @examples
#' res <- KEGGlinks(....)


keggLinks <- function(topGenes, pathwayidlist=NULL, fdr=FALSE, cutoff = 0.05, species = "Hs", foldChange=FALSE, 
			    path_to_write=NULL, sessionID=NULL, debuging=FALSE) {
  tg <- TRUE
  if(species == "hsa") {
    species <- "Hs"
  } else if (species == "mmu") {
    species <- "Mm"
  } else if (species == "rno") {
    species <- "Rn"
  }
  keggPaths <- get(data(list=paste0("keggPathways", species)))
  if(is.null(pathwayidlist)) pathwayidlist <- get(data("keggPathways"))[[species]]
  
  if(!is.data.frame(topGenes)) {
    tg <- FALSE
    
    if(length(na.omit(as.integer(topGenes))) < length(topGenes)/2) {
      data(geneinfo)
      topGenes <- na.omit(geneinfo[match(topGenes, geneinfo$Symbol),"GeneID"])
    }
  if(species == "hsa") {
    species <- "Hs"
  } else if (species == "mmu") {
    species <- "Mm"
  } else if (species == "rno") {
    species <- "Rn"
  }
    homoloArr <- getHomologousGenes(as.character(topGenes), org=species)
    topGenes <- data.frame(geneid=homoloArr$homoloGeneID[!is.na(homoloArr$homoloGeneID)], log2FoldChange=1, pval=0.05, stringsAsFactors=F)
    rownames(topGenes) <- topGenes$geneid
  }
  
  if (foldChange) {
    pvals <- -abs(topGenes$log2FoldChange)
    cutoff <- -log2(cutoff)
  } else if (fdr) {
    pvals <- p.adjust(topGenes$pval, method="fdr")
  } else {
    pvals <- topGenes$pval
  }
  na.pval <- is.na(pvals)
  pvals <- pvals[!na.pval]
  geneid <- rownames(topGenes)[!na.pval]
#  geneid <- topGenes$gid
 names(pvals) <- geneid
  log2FC <- topGenes[geneid, "log2FoldChange"]
  names(log2FC) <- geneid

  
  DEcol <- "red"; notDEcol <- "black"; notindataFG <- "gray"; notindataBG <- "white"
##  library(gplots)
  log2FCthresh <- 1; nGradient <- 5; hiBGcol <- "yellow"; lowBGcol <- "blue"; midBGcol <- "#E6E6E6"  # soft gray
  
  logFCgradient <- gplots::colorpanel(nGradient, low=lowBGcol,mid=midBGcol, high=hiBGcol)
  # Substitute "%23" instead of "#"
  logFCgradient <- gsub("#", "%23", logFCgradient)

  # Store foreground and background colors for all genes
  genecol <- data.frame(geneid=geneid, row.names=geneid)

  # Map foreground colors for differentially expressed or not differentially expressed genes
  genecol$fg <- ifelse(pvals <= cutoff, DEcol, notDEcol)

  index <- pmax(1, pmin(nGradient, floor(1 + (log2FC + log2FCthresh) * (nGradient - 2) / (2 * log2FCthresh)) + 1))
  genecol$bg <- logFCgradient[index]

  if(species == "Hs") {
    species <- "hsa"
  } else if (species == "Mm") {
    species <- "mmu"
  } else if (species == "Rn") {
    species <- "rno"
  }
  if(tg & all(substr(pathwayidlist, 1, 3) != species)) {
    stop("Wrong pathwayidlist format!")
#     pathwayidlist <- paste0(species, pathwayidlist)
  }

###################################
  
  if(!tg) {
    
    genesIn <- lapply(keggPaths, function(j) j[[1]]$GENE[j[[1]]$GENE %in% as.character(topGenes$geneid)])
    res <- data.frame(Nam = gsub(" - Homo sapiens [(]human[)]", "",pathwayidlist),
		      ID = names(pathwayidlist), 
		      pSize = sapply(keggPaths, function(j) length(j[[1]]$GENE)/2), 
		      NDE = sapply(genesIn, function(j) length(j)),# sapply(keggPaths, function(j) sum(j[[1]]$GENE %in% as.character(topGenes$geneid))), 
		      KEGGLINK = NA, 
		      genesIN = NA, # sapply(genesIn, function(j) paste(geneid2symbol(paste(j, collapse=","), debuging = debuging)$Symbol, collapse=",")), # paste(geneid2symbol(paste(overlapIDs, collapse=","), debuging = debuging)$Symbol, collapse=","),
		      Remark = "Done", stringsAsFactors = FALSE)
    res <- res[res$NDE > 0, ]
    if(dim(res)[1] ==0) {
	res[1,] <- NA
	res$Remark <- "No pathway found!"
	return(res)
    }
    genesIn <- genesIn[res$ID]
    res$genesIN <- sapply(genesIn, function(j) paste(geneid2symbol(paste(j, collapse=","), debuging = debuging)$Symbol, collapse=","))
    for(j in 1:nrow(res)) {
	urlbegin <- paste0("http://www.kegg.jp/kegg-bin/show_pathway?map=", res$ID[j], "&multi_query=")
	newlink <- urlbegin
	# Create link text for first gene, then loop through the remaining genes with "%0D%0A" between each one
	gtext <- paste0(genesIn[[j]][1], "+", "grey", "%2C", "red")
	newlink <- paste0(newlink, gtext)
	for(gg in genesIn[[j]][-1]) {
	    if(length(genesIn[[j]]) > 1){
		gtext <- paste0(gg[[1]], "+", "grey", "%2C", "red")
		newlink <- paste0(newlink, "%0D%0A", gtext)
	    }
	}
	res$KEGGLINK[j] <- newlink
    }
    res <- res[order(res$NDE, decreasing=TRUE),]
    colnames(res) <- c("KEGG pathway name", "KEGG pathway ID", "Genes in Pathway", "Found Genes in Pathway", "KEGG link", "Found Genes Names", "Remark")
    return(res)
  }
  
  keggPaths <- keggPaths[pathwayidlist]
  KEGGlinks <- NULL
  genesIN <- NULL
  for(pathway in pathwayidlist) {
#    kegg.path <- KEGGREST::keggGet(paste0("path:", pathway))[[1]]
    print(pathway)
    whichPath <- which(sapply(keggPaths, function(i) i[[1]]$ENTRY)==pathway)
    keggPath <- keggPaths[[whichPath]][[1]]
    pathGenes <- keggPath$GENE
    pathGenesData <- data.frame(matrix(pathGenes, ncol=2, byrow=TRUE), stringsAsFactors=FALSE)
    colnames(pathGenesData) <- c("geneid", "description")
    pathGenesData$geneid <- as.character(pathGenesData$geneid)
    rownames(pathGenesData) <- pathGenesData$geneid
    pathGenesData$keggid <- sub(".*?KO:(.*?)(].*|$)", "\\1", pathGenesData$description)
    keggid.table <- table(pathGenesData$keggid)
    pathGenesData$duplicate <- keggid.table[pathGenesData$keggid] > 1
    
    uniqueIDs <- pathGenesData$geneid[!pathGenesData$duplicate]
    overlapIDs <- intersect(uniqueIDs, geneid)
    missingIDs <- uniqueIDs[!(uniqueIDs %in% geneid)]
    
    pathwayGeneColor <- genecol[overlapIDs, ]
    if(dim(pathwayGeneColor)[1] > 0) {
	pathwayGeneColor[missingIDs, "fg"] <- notindataFG
	pathwayGeneColor[missingIDs, "bg"] <- notindataBG
	pathwayGeneColor$geneid <- rownames(pathwayGeneColor)
    }
    duplicateIDs <- pathGenesData$geneid[pathGenesData$duplicate]
    duplicateKEGGid <- unique(pathGenesData[duplicateIDs, "keggid"])
    for(dd in duplicateKEGGid) {
      duplicatePathGenes <- pathGenesData[pathGenesData$keggid == dd, "geneid"]
      overlapDuplicateIDs <- intersect(duplicatePathGenes, geneid)
      
      # Create row in pathway gene color table for duplicate genes (matching first gene in list)
      pathwayGeneColor[duplicatePathGenes[1], "geneid"] <- duplicatePathGenes[1]
      
      if(length(overlapDuplicateIDs) == 0) {
        # Genes not in data
        pathwayGeneColor[duplicatePathGenes[1], "fg"] <- notindataFG
        pathwayGeneColor[duplicatePathGenes[1], "bg"] <- notindataBG
      } else {
        if(any(pvals[overlapDuplicateIDs] < cutoff)) {
          pathwayGeneColor[duplicatePathGenes[1], "fg"] <- DEcol
        } else {
          pathwayGeneColor[duplicatePathGenes[1], "fg"] <- notDEcol
        }
        
        # Set background colour for all genes equal to most extreme log2FC across the duplicates
        maxLog2FC <- max(log2FC[overlapDuplicateIDs])
        minLog2FC <- min(log2FC[overlapDuplicateIDs])
        duplicateLog2FC <- ifelse(abs(maxLog2FC) > abs(minLog2FC), maxLog2FC, minLog2FC)
        # Find corresponding log2FC color gradient index
        duplicateIndex <- max(1, min(nGradient, floor(1 + (duplicateLog2FC + log2FCthresh) * (nGradient - 2) / (2 * log2FCthresh)) + 1))
        pathwayGeneColor[duplicatePathGenes[1], "bg"] <- logFCgradient[duplicateIndex]
      }
    }
    
    urlbegin <- paste0("http://www.kegg.jp/kegg-bin/show_pathway?map=", pathway, "&multi_query=")
    newlink <- urlbegin
    gene.text <- paste0(pathwayGeneColor[1, "geneid"], "+", pathwayGeneColor[1, "bg"], "%2C", pathwayGeneColor[1, "fg"])
    newlink <- paste0(newlink, gene.text)
    for(gg in pathwayGeneColor$geneid[-1]) {
      gene.text <- paste0(gg, "+", pathwayGeneColor[gg, "bg"], "%2C", pathwayGeneColor[gg, "fg"])
      newlink <- paste0(newlink, "%0D%0A", gene.text)
    }
    KEGGlinks <- c(KEGGlinks, newlink)
    genesIN <- c(genesIN, ifelse(length(overlapIDs) > 0, paste(geneid2symbol(paste(overlapIDs, collapse=","), debuging = debuging)$Symbol, collapse=","), ""))
#  return(newlink)
  }
  
  if(!tg) {
    if(dim(res)[1] ==0) {
	res[1,] <- NA
	res$Remark <- "No pathway found!"
    } else {
	res$KEGGLINK <- KEGGlinks
	res$genesIN <- genesIN
	res$Remark <- "Done"
    }
    res <- res[order(res$NDE, decreasing=TRUE),]
    colnames(res) <- c("KEGG pathway name", "KEGG pathway ID", "Genes in Pathway", "Found Genes in Pathway", "KEGG link", "Found Genes Names", "Remark")
    return(res) 
  } else return(KEGGlinks)
}

