
#' A function for ...
#'
#' This function allows you to ...
#' @param query.p ...
#' @param reference.p ...
#' @param query.gL ...
#' @param reference.gL ...
#' @param na.rm ...
#' @param estimateNullDistr ...
#' @param nullDistrQuantiles ...
#' @param nullDistrN ...
#' @param tolerateWarnings ...
#' @param pAdjust.method.query ...
#' @param pAdjust.method.reference ...
#' @param lambda ...
#' @param rescale ...
#' @param plotRescaled ...
#' @param multicoreN ...
#' @param authors M. Fazel-Najafabadi
#' @keywords ...
#' @export 
#' @examples
#' res <- batchTregGRS(....)

nCols_equals_three <- function(sessionID,sortedValues,layoutMatrix,allColors,pValue,prop,nFactorLevels,factorLevels,nColors,nObs,sortedFactors,path_to_write)
{
	pdf(paste(path_to_write,"temptt",sessionID,".pdf",sep=""))
	yLimits <- range(sortedValues,na.rm=T)
	layout(layoutMatrix,heights=c(1,5),widths=5)
	par(mar=c(4,5,2,5))
	plot(1:nObs,as.numeric(sortedValues),pch=19,col=allColors[1],xlab="Samples",ylab="Gene Expression Level",ylim=yLimits,
	main=paste("p-value for differential expression =",round(pValue,4),"\n for different levels of factor ",prop))

	for(i in 2:nFactorLevels) {
		points((1:nObs)[sortedFactors==factorLevels[i]],as.numeric(sortedValues[sortedFactors==factorLevels[i]]),pch=19,col=allColors[((i-1)%%nColors)+1])
	}
	par(mar=c(0,5,0,5),bty="n")
	plot(c(0,1),c(0,1),type="n",xaxt="n",yaxt="n",xlab="",ylab="")
	legend(x=0,y=1,legend=factorLevels,col=allColors[(((1:nFactorLevels)-1)%%nColors)+1],pch=rep(19,nFactorLevels),ncol=3)
	try(dev.off())

}
